File: hosts/websocket/csharp/Discord.cs
using System;
using System.Linq;
using System.Runtime.InteropServices;
using System.Net.Sockets;
using System.Collections.Generic;
using System.IO;
using System.IO.Pipes;
using System.Threading;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;
using System.Security.Principal;
using System.Diagnostics;


public record ActivityTimestamps(
	[property: JsonPropertyName("start")] long? Start,
	[property: JsonPropertyName("end")] long? End
);

public record ActivityAssets(
	[property: JsonPropertyName("large_image")] string? LargeImage,
	[property: JsonPropertyName("large_text")] string? LargeText,
	[property: JsonPropertyName("small_image")] string? SmallImage,
	[property: JsonPropertyName("small_text")] string? SmallText
);

public record ActivityButton(
	[property: JsonPropertyName("label")] string Label,
	[property: JsonPropertyName("url")] string Url
);

public record Activity(
	[property: JsonPropertyName("state")] string State,
	[property: JsonPropertyName("details")] string Details,
	[property: JsonPropertyName("timestamps")] ActivityTimestamps? Timestamps,
	[property: JsonPropertyName("assets")] ActivityAssets Assets,
	[property: JsonPropertyName("buttons")] ActivityButton[]? Buttons
);

public record ResponseActivityMetadata(
	[property: JsonPropertyName("button_urls")] string[] ButtonURLs
);

public record ResponseActivity(
	[property: JsonPropertyName("state")] string State,
	[property: JsonPropertyName("details")] string Details,
	[property: JsonPropertyName("timestamps")] ActivityTimestamps? Timestamps,
	[property: JsonPropertyName("assets")] ActivityAssets Assets,
	[property: JsonPropertyName("buttons")] string[] Buttons,
	[property: JsonPropertyName("name")] string Name,
	[property: JsonPropertyName("application_id")] string ApplicationId,
	[property: JsonPropertyName("type")] int type,
	[property: JsonPropertyName("metadata")] ResponseActivityMetadata Metadata
);

public record Error(int code, string message);

public record SetActivityResponse<T>(
	string? cmd,
	T data,
	string? evt,
	string? nonce
);

public record AuthorizationResponseDataConfig(
	string cdn_host,
	string api_endpoint,
	string environment
);

public record AuthorizationResponseDataUser(
	string id,
	string username,
	string discriminator,
	string global_name,
	string avatar,
	string? avatar_decoration_data,
	bool bot,
	int flags,
	int premium_type
);

public record AuthorizationResponseData(
	int v,
	AuthorizationResponseDataConfig config,
	AuthorizationResponseDataUser user
);

public record AuthorizationResponse(
	string cmd,
	AuthorizationResponseData data,
	string evt,
	string? nonce
);


public class Discord {
	StreamString? buf = null;
	NamedPipeClientStream? pipe = null;
	Socket? socket = null;
	string os = "Unsupported";

	public Discord() {
		if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
			os = "Windows";
		else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
			os = "Linux";
		else
			Console.WriteLine("Unsupported OS");
	}

	public (bool connected, string[] errors) Connect() {
		if (os == "Unsupported") return (false, new string[] {"Unsupported OS"});

		string[] sockets = this.GetSockets();
		var errors = new string[sockets.Length > 0 ? sockets.Length : 1];

		if (sockets.Length == 0)
			errors[0] = "No sockets running";

		for (int i = 0; i < sockets.Length; ++i)
			switch (this.os) {
				case "Windows": {
					this.pipe = new NamedPipeClientStream(
						".", sockets[i],
						PipeDirection.InOut,
						PipeOptions.None,
						TokenImpersonationLevel.Impersonation
					);

					try { this.pipe.Connect(); }
					catch (Exception e) { errors[i] = e.Message; continue; }

					this.buf = new StreamString(this.pipe);
					return (true, errors);
				}

				case "Linux": {
					this.socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
					var endPoint = new UnixDomainSocketEndPoint(sockets[i]);

					try { this.socket.Connect(endPoint); }
					catch (SocketException e) { errors[i] = e.Message; continue; }
					catch (Exception e) { errors[i] = e.Message; continue; }

					this.buf = new StreamString(this.socket);
					return (true, errors);
				}
			}

		return (false, errors);
	}

	string[] GetSockets() {
		string filename, args;

		switch (this.os) {
			case "Windows":
				filename = "powershell";
				args = @"-Command (Get-ChildItem \\.\pipe\).Name | findstr discord";
				break;

			case "Linux":
				filename = "/bin/sh";
				args = "-c \"ss -lx | grep -o '[^[:space:]]*discord[^[:space:]]*'\"";
				break;

			default:
				return new string[] {};
		}

		var startInfo = new ProcessStartInfo() {
			FileName = filename,
			Arguments = args,
			RedirectStandardOutput = true,
			RedirectStandardError = true,
			UseShellExecute = false,
			CreateNoWindow = true
		};

		using (Process process = Process.Start(startInfo)!) {
			string stdout = process!.StandardOutput.ReadToEnd();
			// string stderr = process!.StandardError.ReadToEnd();

			return stdout
				.Split("\n")
				.Select(line => line.Trim())
				.Where(line => line.Length > 0)
				.ToArray();
		}
	}

	public (bool success, string? msg, object? res) Authorize() {
		if (os == "Unsupported") return (false, "Unsupported OS", null);

		var payload = new {
			client_id = "1059272441194623126",
			v = 1
		};

		var msg = Call(0, payload);
		var err = JsonSerializer.Deserialize<Error>(msg);

		if (err?.code > 0)
			return (false, err.message, null);

		var res = JsonSerializer.Deserialize<AuthorizationResponse>(msg);
		return (true, null, res);
	}

	public (bool success, string? msg, object? res) ClearActivity() {
		if (os == "Unsupported") return (false, "Unsupported OS", null);

		return this.SetActivity(null);
	}

	public (bool success, string? msg, object? res) SetActivity(Activity? activity) {
		if (os == "Unsupported") return (false, "Unsupported OS", null);

		var process = Process.GetCurrentProcess();

		var payload = new {
			cmd = "SET_ACTIVITY",
			nonce = Guid.NewGuid().ToString(),
			args = new {
				activity,
				pid = process.Id
			}
		};

		var msg = Call(1, payload);
		var err = JsonSerializer.Deserialize<Error>(msg);

		if (err?.code > 0)
			return (false, err.message, null);

		var data = JsonSerializer.Deserialize<SetActivityResponse<Error>>(msg);

		if (data?.evt == "ERROR")
			return (false, data?.data.message, null);

		var res = JsonSerializer.Deserialize<SetActivityResponse<ResponseActivity>>(msg);
		return (true, msg, res);
	}

	string Call(int opcode, object payload) {
		var options = new JsonSerializerOptions {
			DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,

			// Discord requires indentation, otherwise:
			//     Unterminated string in JSON at position x (line 1 column x)
			WriteIndented = true
		};

		string packed_payload = JsonSerializer.Serialize(payload, options);
		byte[] packed_opcode = BitConverter.GetBytes(opcode);
		byte[] packed_size = BitConverter.GetBytes(packed_payload.Length);

		string data = string.Concat(
			Encoding.ASCII.GetString(packed_opcode),
			Encoding.ASCII.GetString(packed_size),
			packed_payload
		);
		this.buf!.WriteString(data);

		byte[] res = this.buf.ReadBytes();
		return Encoding.ASCII.GetString(res);
	}
}

class Printing {
	static void PrintByteArray(string byteArray) {
		foreach (char b in byteArray)
				if (Printing.IsPrintable(b)) Console.Write($"{b} ");
				else Console.Write($"{(int) b:X2} ");
		Console.WriteLine();
	}

	static void PrintByteArray(byte[] byteArray) {
		foreach (byte b in byteArray)
				if (Printing.IsPrintable(b)) Console.Write($"{(char) b} ");
				else Console.Write($"{b:X2} ");
		Console.WriteLine();
	}

	static bool IsPrintable(char b) {
		return !char.IsControl(b) && b >= 32 && b <= 126;
	}

	static bool IsPrintable(byte b) {
		return b >= 32 && b <= 126;
	}
}

class StreamString {
	Stream ioStream;
	Socket socket;
	string os;

	public StreamString(Stream ioStream) {
		this.os = "Windows";
		this.ioStream = ioStream;
	}

	public StreamString(Socket socket) {
		this.os = "Linux";
		this.socket = socket;
	}

	public byte[] ReadBytes() {
		var metadataBytes = new byte[8];

		switch (this.os) {
			case "Windows": ioStream.Read(metadataBytes, 0, metadataBytes.Length); break;
			case "Linux": socket.Receive(metadataBytes); break;
		}

		var size = BitConverter.ToInt32(metadataBytes[4..], 0);
		var inBuffer = new byte[size];

		switch (this.os) {
			case "Windows": ioStream.Read(inBuffer, 0, size); break;
			case "Linux": socket.Receive(inBuffer); break;
		}

		return inBuffer;
	}

	public void WriteString(string outString) {
		byte[] outBuffer = Encoding.ASCII.GetBytes(outString);

		switch (this.os) {
			case "Windows":
				ioStream.Write(outBuffer, 0, outBuffer.Length);
				ioStream.Flush();
				break;

			case "Linux": socket.Send(outBuffer); break;
		}
	}
}

File: hosts/websocket/python/discord.py
import os
import sys
import time
import subprocess
import json
import struct
import uuid
from typing import TypedDict, Optional, Any
from enum import Enum

if sys.platform == 'linux':
	import socket
elif sys.platform == 'win32':
	import win32file
	import win32pipe


# Discord data structures
class ActivityTimestamps(TypedDict, total=False):
	start: Optional[int]
	end: Optional[int]

class ActivityAssets(TypedDict, total=False):
	large_image: Optional[str]
	large_text: Optional[str]
	small_image: Optional[str]
	small_text: Optional[str]

class ActivityButton(TypedDict):
	label: str
	url: str

class Activity(TypedDict, total=False):
	state: str
	details: str
	timestamps: Optional[ActivityTimestamps]
	assets: ActivityAssets
	buttons: Optional[list[ActivityButton]]

class ResponseActivityMetadata(TypedDict):
	button_urls: list[str]

class ResponseActivity(TypedDict):
	state: str
	details: str
	timestamps: Optional[ActivityTimestamps]
	assets: ActivityAssets
	buttons: list[str]
	name: str
	application_id: str
	type: int
	metadata: ResponseActivityMetadata

class Error(TypedDict):
	code: int
	message: str

class SetActivityResponse(TypedDict):
	cmd: str
	data: Activity | Error
	evt: str
	nonce: str

class AuthorizationResponseDataConfig(TypedDict):
	cdn_host: str
	api_endpoint: str
	environment: str

class AuthorizationResponseDataUser(TypedDict):
	id: str
	username: str
	discriminator: str
	global_name: str
	avatar: str
	avatar_decoration_data: Optional[str]
	bot: bool
	flags: int
	premium_type: int

class AuthorizationResponseData(TypedDict):
	v: int
	config: AuthorizationResponseDataConfig
	user: AuthorizationResponseDataUser

class AuthorizationResponse(TypedDict):
	cmd: str
	data: AuthorizationResponseData
	evt: str
	nonce: Optional[str]

# Library data structures
class ConnectStatus(Enum):
	Connected = 0
	SocketsRetriveError = 1
	NoSockets = 2
	ConnectionError = 3


class StreamString:
	os:     str
	socket: Any

	def __init__(self, os: str, socket: Any):
		self.os = os
		self.socket = socket

	def read_bytes(self) -> tuple[bool, str|None, tuple[int, bytes]|None]:
		try:
			res = self.read_bytes_()
			return (True, None, res)
		except Exception as e:
			return (False, str(e), None)

	def write_bytes(self, data: bytes) -> tuple[bool, str|None]:
		try:
			self.write_bytes_(data)
			return (True, None)
		except Exception as e:
			return (False, str(e))

	def close(self):
		match self.os:
			case 'linux': self.socket.close()
			case 'win32': win32file.CloseHandle(self.socket)

	def read_bytes_(self) -> tuple[int, bytes]:
		raw_metadata: bytes
		buf: bytes

		match self.os:
			case 'linux': raw_metadata = self.socket.recv(8)
			case 'win32': raw_metadata = win32file.ReadFile(self.socket, 8)[1]
			case _: raise Exception('Unsupported system')

		metadata = struct.unpack('<II', raw_metadata)
		size = metadata[1]

		match self.os:
			case 'linux': buf = self.socket.recv(size)
			case 'win32': buf = win32file.ReadFile(self.socket, size)[1]
			case _: raise Exception('Unsupported system')

		return (metadata[0], buf)

	def write_bytes_(self, data: bytes):
		match self.os:
			case 'linux': self.socket.sendall(data)
			case 'win32': win32file.WriteFile(self.socket, data)
			case _: raise Exception('Unsupported system')


class Discord:
	client_id: str
	connected: bool
	os:        str

	socket:    Any
	buf:       StreamString

	def __init__(self, client_id: str):
		self.client_id = client_id
		self.connected = False
		self.os = sys.platform

		if self.os not in ('linux', 'win32'):
			raise Exception('Not supported system')

	def authorize(self) -> tuple[bool, str|None, AuthorizationResponse|Error|None]:
		if not self.connected: return (False, 'Not connected', None)

		payload = {
			'client_id': self.client_id,
			'v': 1
		}

		done, msg, res = self.__call(0, payload)
		if not done:
			self.connected = False
			return (False, msg, None)
		if not res: return (False, msg, None) # Remove |None from res

		_, msg = res
		data = json.loads(msg)

		if 'code' in data: return (False, msg, Error(data))

		return (True, msg, AuthorizationResponse(data))

	def clear_activity(self):
		return self.set_activity(None)

	def set_activity(self, activity: Activity|None) -> tuple[bool, str|None, SetActivityResponse|Error|None]:
		if not self.connected: return (False, 'Not connected', None)

		payload = {
			'cmd': 'SET_ACTIVITY',
			'nonce': str(uuid.uuid4()),
			'args': {
				'activity': activity,
				'pid': os.getpid()
			}
		}

		done, msg, res = self.__call(1, payload)
		if not done:
			self.connected = False
			return (False, msg, None)
		if not res: return (False, msg, None) # Remove |None from res

		_, msg = res
		data = json.loads(msg)

		if 'code' in data: return (False, msg, Error(data))
		if data.get('evt') == 'ERROR': return (False, msg, SetActivityResponse(data))

		return (True, msg, SetActivityResponse(data))

	def connect(self) -> tuple[ConnectStatus, str|None]:
		if self.connected: self.buf.close()

		done, msg, sockets = self.__get_sockets()
		if not done: return (ConnectStatus.SocketsRetriveError, msg)
		if not sockets: return (ConnectStatus.NoSockets, None)

		match self.os:
			case 'linux': self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)

		for sock in sockets:
			try:
				match self.os:
					case 'linux':
						self.socket.connect(sock)
						self.socket.setblocking(False)

					case 'win32':
						self.socket = win32file.CreateFile(
							sock,
							win32file.GENERIC_READ | win32file.GENERIC_WRITE,
							0, None,
							win32file.OPEN_EXISTING,
							0, None
						)

				self.buf = StreamString(self.os, self.socket)
				self.connected = True
				return (ConnectStatus.Connected, None)
			except: ...

		return (ConnectStatus.ConnectionError, "Couldn't connect to any socket")

	def __get_sockets(self) -> tuple[bool, str|None, list[str]|None]:
		cmd: str

		match self.os:
			case 'linux': cmd = 'ss -lx | grep -o [^[:space:]]*discord[^[:space:]]*'
			case 'win32': cmd = 'powershell -Command (Get-ChildItem \\\\.\\pipe\\).FullName | findstr discord'
			case _: raise Exception('Unsupported system')

		process = subprocess.run(cmd, capture_output=True, text=True, shell=True)
		stdout = process.stdout
		stderr = process.stderr

		if stderr:
			return (False, stderr, None)

		return (True, None, stdout.strip().split('\n'))

	# might throw error
	def __call(self, opcode: int, payload: dict) -> tuple[bool, str|None, tuple[int, str]|None]:
		data = json.dumps(payload)
		msg = struct.pack('<ii', opcode, len(data)) + data.encode()

		done, msg = self.buf.write_bytes(msg)
		if not done: return (False, msg, None)

		time.sleep(2)

		done, msg, res = self.buf.read_bytes()
		if not done: return (False, msg, None)
		if not res: return (False, msg, None)

		opcode = res[0]
		data = res[1].decode()

		return (True, None, (opcode, data))

				# res = b''
				#
				# while True:
				# 	try:
				# 		chunk = self.socket.recv(1024)
				# 		if not chunk: break
				# 		res += chunk
				# 	except BlockingIOError:
				# 		break
				# 	except socket.timeout:
				# 		# Timeout occurred while waiting for data
				# 		break
				#
				# opcode = struct.unpack('<ii', res[:8])[0]
				# msg = res[8:].decode()

if __name__ == '__main__':
	discord = Discord('1059272441194623126')
	connected = discord.connect()

	if not connected: print('Not connected')
	else:
		print('Connected')

		authorized = discord.authorize()
		if not authorized: print('Not authorized')
		else:
			print('Authorized')

			activity: Activity = {
				'state': 'Some state',
				'details': 'Some details',
				'assets': {
					'large_image': 'https://logowik.com/content/uploads/images/t_283_wikipedia.jpg'
				}
			}

			discord.set_activity(activity)
			print('Set activity')
			input()

			discord.clear_activity()
			print('Set activity')
			input()
